### PR TITLE
Add flag to decide whether we should call `git fetch` before operations

### PR DESCRIPTION
**Pull request by mujx**
## Description

Add flag to decide whether we should call `git fetch` before operations.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/ops/cli/parser.py [hunk 1]
```diff
@@ -34,6 +34,18 @@ def _get_parser(self):
...
+        gitfetchHelp = "Whether to call `git fetch` and receive updates from the infrastructure repository." \
+                       " To be used in conjunction with the terraform or helmfile commands"
+        git_fetch_parser = parser.add_mutually_exclusive_group(required=False)
+        git_fetch_parser.add_argument(
+            '--git-fetch', dest='git_fetch', action='store_true', help=gitfetchHelp
+        )
+        git_fetch_parser.add_argument(
+            '--no-git-fetch', dest='git_fetch', action='store_false', help=gitfetchHelp
+        )
+        parser.set_defaults(git_fetch=True)
```
Comment (titirigaiulian) [inline src/ops/cli/parser.py]: Maybe we can keep just one arg and call it more generic like `git_remote` or similar to filter git ops that require interaction with origin

Comment (mujx) [reply src/ops/cli/parser.py]: How the call will look like?  `ops --git-remote <?>` ?

**Comment by mujx** [comment]
@amuraru This is pretty much the current setup. 

We have these two parameters. One for the upstream which can be a URI compatible with the git library and another for specifying where to clone the repo in the filesystem.

The only difference is `file://` which I'm not convinced has a purpose, cause you'll essentially copy the repo from one directory to another. 

The version of infra repo is specified in the hierarchical data, because different projects might use different versions so you'll have to update `.opsconfig` every time you perform operations on different clusters.

**Comment by amuraru** [comment]
> The version of infra repo is specified in the hierarchical data, because different projects might use different versions so you'll have to update .opsconfig every time you perform operations on different clusters.

Understood, makes sense

>We have these two parameters. One for the upstream which can be a URI compatible with the git library and another for specifying where to clone the repo in the filesystem.

That's my point - currently you're assuming git only. I was proposing a _scheme_ based origin and support other sources than git (e.g. Artifactory, Plain URLs, etc)



I still have a strong opinion on this PR change though, so -1 until we change it to be more clear.

You're basically saying:

If opsconfig is:
```
*.upstream_location: "git://git@git.corp.adobe.com/experience-edge/ee-k8s-infra@master"
*.local_cache_dir: ~/.ops/modules/<modulename>
```
and I pass:
`--git-no-fetch`

then `.upstream_location` is totally ignored and whatever I have in `local_cache_dir` is used.
What if `local_cache_dir` dir is empty? Or it's not a precloned dir (we do this internally, I know).
There will be an ugly error that you cannot checkout to `master` branch. 
The behaviour is definetly confusing.

If we really want to support both git clone/fetch/checkout + local modules we need to find a better way.

**Comment by mujx** [comment]
The flag `--no-git-fetch` is used to skip `git fetch` on an already cloned repo. `git fetch` is necessary in case the specified git ref is not available locally. But this can only be done if we have access to the git server at that time.

The flag doesn't interfere with with `upstream_location`. If nothing is found in the `local_cache_dir` the repo will be cloned, otherwise we proceed with `git fetch` & `git checkout <version>`.

The current implementation supports only versioning with a git repo. Handling more sources seems to be beyond the scope of this PR.

File: src/ops/cli/terraform.py [hunk 2]
```diff
@@ -259,7 +259,8 @@ def run_v2_composition(self, args, terraform_path, composition):
         checkout_repo(
```
Comment (amuraru) [inline src/ops/cli/terraform.py]: add a log message before saying what it's happening below.

File: src/ops/git_utils.py [hunk 3]
```diff
@@ -31,14 +31,16 @@ def setup_repo(repo_path, upstream_repo):
...
-def checkout_repo(repo_path, config_path, get_version):
+def checkout_repo(repo_path, config_path, get_version, git_fetch=True):
     with open(os.path.expanduser(config_path)) as f:
         conf = yaml.load(f, Loader=yaml.FullLoader)
 
     version = get_version(conf)
     repo = git.Repo(repo_path, search_parent_directories=True)
 
-    repo.git.fetch()
+    if git_fetch:
```
Comment (amuraru) [inline src/ops/git_utils.py]: I think the `git fetch` should be moved to `setup_repo` method and let `checkout_repo` do only checkout
This way we keep all actions that involve a remote git origin in a single place

**Comment by amuraru** [comment]
To make progress here, @mujx I suggest 

1.  Move `git fetch` part in `setup_repo` - so we have all git actions that involve a remote git origin in that method. 

2. `git-[no]-fetch` flag is not needed. Users (e.g spinnaker) that cannot access git should:
*  pre-clone the repos in `terraform.root_path` and `helmfile.root_path` 
* *ensure* `terraform.upstream_repo` and `helmfile.upstream_repo` **are not defined** in `.opsconfig.yaml` 

3. Please also add logging pre git operations to inform the users what's happening

4. Ideally, handle git errors and report better error messages

5. There is one behaviour that wanted to understand if it's on purpose or a bug

5.1   Define . 
```
terraform.upstream_repo: "git@git.corp.adobe.com/experience-edge/ee-k8s-infra"
terraform.root_path: ../ee-k8s-infra
```
5.2 . My `../ee-k8s-infra` already has the git code checked out in a `master`  branch but the local `master` branch is behind the `origin/master` with 2 commits
`master 232ba44 [origin/master: behind 2]`

When I run `ops ... terraform` the local `master` branch (stale) is used and not the origin.
Is this on purpose? Should we at least fire a WARN message notifying the user it's using an older version of `master` ?

File: src/ops/cli/terraform.py (hunk 2 above)
Comment (mujx) [reply src/ops/cli/terraform.py]: There is already logging inside the function.

**Comment by mujx** [comment]
If you are using a ref that will be updated, you'll need `git pull` to merge the updates. This will have to be handled manually in most cases (in case you have changes locally etc). Probably a flag that will toggle the `git pull` operation might be a good idea.

I'm starting to think that we're reinventing the wheel regarding version management, and we should probably have a proper package manager handle all this. [nix](https://nixos.org/nix/) comes to mind as a good option.

**Comment by titirigaiulian** [comment]
@amuraru @mujx How will it look the run without a flag? As far as I know, we have opsconfig file in our "configs" repo with upstream_repo setup (you want to make sure you run with latest changes locally), but running this in a CI/CD pipeline(inside a docker) will be messy to update the opsconfig file. Having a feature flag and let the user decide at runtime if he wants to run with his local version or to pull latest changes doesn't seem a bad idea after all.
As a long shot we can also override values from the cmd, like the helm cmd
`helm install chart --set key=value`

**Comment by mujx** [comment]
Closing because we switched up the way we do versioning. I would suggest to revert the previous versioning commits also.

**Comment by amuraru** [comment]
> Closing because we switched up the way we do versioning. 

Can you share more?

> I would suggest to revert the previous versioning commits also.

Please do, yes

**Pull request closed without merging**